packages/grafana-ui/src/components/uPlot/Plot.tsx

```diff
@@ -0,0 +1,119 @@
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { css } from 'emotion';
+import uPlot from 'uplot';
+
+import { useTheme } from '../../themes';
+import { buildPlotContext, PlotContext, PlotPluginsContext } from './context';
+import { buildPlotConfig, pluginLog, preparePlotData } from './utils';
+import { PlotPlugin, PlotProps } from './types';
+
+// uPlot abstraction responsible for plot initialisation, setup and refresh
+// Receives a data frame that is x-axis aligned, as of https://github.com/leeoniya/uPlot/tree/master/docs#data-format
+// Exposes contexts for plugins registration and uPlot instance access
+export const UPlotChart: React.FC<PlotProps> = props => {
+  const theme = useTheme();
+  const canvasRef = useRef<HTMLDivElement>(null);
+
+  // instance of uPlot, exposed via PlotContext
+  const [plotInstance, setPlotInstance] = useState<uPlot>();
+
+  // Array with current plot data points, calculated when data frame is passed to a plot
+  const [plotData, setPlotData] = useState<uPlot.AlignedData>();
+  // uPlot config
+  const [config, setConfig] = useState<uPlot.Options>();
+  // map of registered plugins (via children);
+  const [plugins, setPlugins] = useState<Record<string, PlotPlugin>>({});
+
+  const registerPlugin = useCallback(
+    (plugin: PlotPlugin) => {
+      pluginLog(plugin.id, false, 'register');
+      if (plugins.hasOwnProperty(plugin.id)) {
+        throw new Error(`${plugin.id} that is already registered`);
+      }
+      setPlugins(plugs => {
+        return {
+          ...plugs,
+          [plugin.id]: plugin,
+        };
+      });
+
+      return () => {
+        setPlugins(plugs => {
+          pluginLog(plugin.id, false, 'unregister');
+          delete plugs[plugin.id];
+          return {
+            ...plugs,
+          };
+        });
+      };
+    },
+    [setPlugins]
+  );
+
+  useEffect(() => {
+    // Creates array of datapoints to be consumed by uPLot
+    const data = preparePlotData(props.data);
+    // Creates series, axes and scales config
+    // TODO: use field config
+    const config = buildPlotConfig(props, props.data, theme);
+
+    config.plugins = [
+      ...config.plugins,
+      ...Object.entries(plugins).map(p => ({
+        hooks: p[1].hooks,
+      })),
+    ];
+
+    setConfig(config);
+    setPlotData(data);
+  }, [props.data, plugins]);
```

Comment:
I think I found some more deps
```suggestion
  }, [plugins, setConfig, setPlotData, props]);
```